tsl/src/gapfill/planner.c

```diff
@@ -336,8 +425,82 @@ plan_add_gapfill(PlannerInfo *root, RelOptInfo *group_rel)
 
 		foreach (lc, copy)
 		{
-			add_path(group_rel, gapfill_path_create(root, lfirst(lc), context.call));
+			add_path(group_rel, gapfill_path_create(root, lfirst(lc), context.call.func));
 		}
 		list_free(copy);
 	}
 }
+
+static inline bool
+is_gapfill_path(Path *path)
+{
+	return IsA(path, CustomPath) && castNode(CustomPath, path)->methods == &gapfill_path_methods;
+}
+
+/*
+ * Since we construct the targetlist for the gapfill node from the
+ * final targetlist we need to adjust any intermediate targetlists
+ * between toplevel window agg node and gapfill node
+ */
+void
+gapfill_adjust_window_targetlist(PlannerInfo *root, RelOptInfo *input_rel, RelOptInfo *output_rel)
+{
+	ListCell *lc;
+
+	if (!is_gapfill_path(linitial(input_rel->pathlist)))
+		return;
+
+	foreach (lc, output_rel->pathlist)
+	{
+		WindowAggPath *toppath = lfirst(lc);
+
+		/*
+		 * if we have multiple window agg level we need to adjust targetlists of non-toplevel
+		 * WindowAggPaths
+		 */
+		if (IsA(toppath, WindowAggPath) && toppath->winclause->winref > 1)
+		{
+			WindowAggPath *path;
+			/*
+			 * we skip toplevel WindowAggPath because targetlist of toplevel WindowAggPath
+			 * is our starting point for building gapfill targetlist so we don't need to
+			 * adjust the toplevel targetlist
+			 */
+			for (path = (WindowAggPath *) toppath->subpath; IsA(path, WindowAggPath);
+				 path = (WindowAggPath *) path->subpath)
+			{
+				List *exprs = NIL;
+				ListCell *lc_expr;
+
+				foreach (lc_expr, toppath->path.pathtarget->exprs)
+				{
+					gapfill_walker_context context;
+
+					gapfill_expression_walker(lfirst(lc_expr), window_function_walker, &context);
+
+					/*
+					 * we error out on multiple window functions per column when building
+					 * gapfill node targetlist so we only assert here
+					 */
+					Assert(context.count <= 1);
+
+					if (context.count == 1)
+					{
+						if (context.call.window->winref <= path->winclause->winref)
+							/*
+							 * window function of current level or below
+							 * so we can put in verbatim
+							 */
+							exprs = lappend(exprs, lfirst(lc_expr));
+						else if (context.call.window->args != NIL &&
+								 contain_var_clause(linitial(context.call.window->args)))
+							exprs = lappend(exprs, linitial(context.call.window->args));
```

Comment:
we should also error if there is a var clause in the non-first argument

Reply:
done